docs/debug/cpu_frequency.md

```diff
@@ -156,11 +156,59 @@ This can be the source of some reboot problems, but it is hard to test in a
 reliable way - reset is asynchronous event that have to take place in very short
 amount of time during transition.
 
+## Voltage regulator
+
+CPU communicates with voltage regulator using SVI2 (Serial VID Interface 2.0).
+Typically, after all frequency and voltage dependencies described above are
+resolved, transition to higher performance follows a sequence:
+
+1. SVI2 command is send to the voltage regulator, `D18F5x12C[Svi2CmdBusy]`
+   is set.
+2. Voltage regulator sets the requested voltage, waits for it to stabilize and
+   clears `Svi2CmdBusy`.
+3. After `Svi2CmdBusy` is clear frequency change is performed.
+
+Transition towards lower performance starts with frequency change and doesn't
+have to wait for it to finish before changing voltage. Because of these
+restrictions it is impossible for CPU to work with high frequency and too low
+voltage for that frequency.
+
+Multiple requests can be grouped:
+
+> * If multiple commands are issued that affect the P-state of a domain prior
+>   to when the processor initiates the change of the P-state of that domain,
+>   then the processor operates on the last one issued.
+> * Once a P-state change starts, the P-state state machine (PSSM) continues
+>   through completion unless interrupted by a PWROK deassertion. If multiple
+>   P-state changes are requested concurrently, the PSSM may group the
+>   associated VID changes separately from the associated COF changes.
+
+Setting `D18F5x12C[WaitVidCompDis]` changes behaviour of transition towards
+higher performance - instead of waiting for voltage regulator to report end of
+transition next request can be made after a defined time period
+(`D18F3xD8[VSRampSlamTime]`).
+
+## Findings from logs obtained from community
+
+Registers values from community revealed that CPU had voltage of (software) P0,
+but its frequency was stuck at a value of P2. Voltage change to P0 was requested
+but not reported as done by voltage regulator for some reason.
```

Comment:
@krystian-hebel this should be goal of some further research - question is what tools we have to narrow this down?

Reply:
@pietrushnic it is (or at least should be) reported through SVI2. There is no publicly available specification for this. Anyway, I doubt there are any test points between voltage converter and CPU.

Search for the specification lead to http://www.zeroplus.com.tw/logic-analyzer_en/news_images/AMD%20SVI2-en.pdf

Reply:
@krystian-hebel there are some parts compatible with SVI2:

* https://docplayer.net/51160536-Rt8870a-dual-output-pwm-controller-for-amd-svi-mobile-cpu-power-supply.html
* https://www.infineon.com/dgdl/Infineon-IR35201MTRPBF-DS-v01_00-EN.pdf?fileId=5546d462576f347501579c95d19772b5
* https://media.digikey.com/pdf/Data%20Sheets/ON%20Semiconductor%20PDFs/NCP81022.pdf

Anyway to not make this discussion longer, please fill JIRA item in content marketing and we will see if there will be time and resources to take care of that in future.